QueryParams: Accept tuples and sets but not dictionaries when setting a key

willhuang1997 opened the pull request:
**NOTE**: This PR has to be merged after https://github.com/streamlit/streamlit/pull/7718 gets merged.

<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
- add an extra check for `tuple` and `set`; if it's one of those types, convert to `list` of `str`
- throw a `StreamlitAPIException` when getting a `dict` type
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
  - added unit tests
- E2E Tests
- Any manual testing needed?
---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

lukasmasuch committed:
Add support for timedelta type to `st.dataframe`, `st.data_editor` and `st.table`. (#7689)

* Add support for timedelta type

* Add tests

* Move timedelta to supported types

* Improve timedelta support

* Improve implementation

* Rename to convert to seconds

* Fix type util test

* Fix unit tests

* Add timedelta index

* Update notices file

* Update snapshots

* Update snapshots

* Apply review comments

kajarenc committed:
Add Python 3.12 support (#7663)

Updates Streamlit for Python 3.12 compatibility

We have a test that checks interaction with langchain (callback handler), which is not yet 3.12-compatible. These are skipped when running on 3.12.
ALL_PYTHON_VERSIONS are extended with 3.12 for our Github Actions runners
Refactor a few tests to use self.assertEqual instead of deprecated  self.assertEqualS
Bump revision of insert-licence pre-commit hook to work with Python 3.12

lukasmasuch committed:
Add border parameter to container and form (#7455)

* Add border parameter to container and form

* Add border parameter

* Add snapshots

* Add note to form docstring

vdonato committed:
Add exhaustive proto stability tests (#7635)

kajarenc committed:
temp fix mypy upper bound, to avoid broken CI (#7713)

lukasmasuch committed:
Enforce pyarrow version for arrow-based custom components (#7695)

lukasmasuch committed:
Release Streamlit version 1.28.2 (#7716)

* Build releases directly from tags instead of release branches (#7580)

* Get rid of unused release candidate workflow

* Just build releases directly from tag instead of release branch

* Get rid of unused scripts

* Add back branch check to verify that corresponding PR exists

* Docstrings for App testing (#7587)

* Make the outline for expanders appear on focus-visible and not focus (#7592)

* Add new libraries to attributions (#7599)

* Add new libraries to attributions

* Update metrics_util.py

* Up version to 1.28.0

* Fix: MPA Nav expand arrow (#7634)

Update dependency array for isOverflowing check - does not always register changes in clientHeight after initial expand & collapse. Add expanded to dependency array to ensure check of scrollHeight vs. clientHeight

* Tweak "forgot config" SnowflakeConnection error message (#7652)

* Fix app testing repr bug for `st.container` (#7644)

* Ensure file_uploader doesn't trigger needless reruns (#7641)

* Fix trigger value regression with `st.rerun` (#7643)

* Greatly narrow errors that we retry in SnowflakeConnection (#7645)

* Fix issue for st.camera_input and chat_input, add regression test for file_uploader (#7646)

Add playwright regression test for issue With st.file_uploader and st.chat_input, first message disappears #7556
Replicate changes from Ensure file_uploader doesn't trigger needless reruns #7641 for st.camera_input

* Up version to 1.28.1

* Add fix to enforce pyarrow version

* temp fix mypy upper bound, to avoid broken CI (#7713)

* Update version to 1.28.2

---------

Co-authored-by: Vincent Donato <vincent@streamlit.io>
Co-authored-by: Debbie Matthews <debbie.matthews@snowflake.com>
Co-authored-by: Ken McGrady <ken.mcgrady@gmail.com>
Co-authored-by: Maya Barnes <63436329+mayagbarnes@users.noreply.github.com>
Co-authored-by: Amanda Walker <amanda@amandawalker.io>
Co-authored-by: Karen Javadyan <kajarenc@gmail.com>

dependabot[bot] committed:
Bump actions/setup-node from 3 to 4 (#7628)

Bumps [actions/setup-node](https://github.com/actions/setup-node) from 3 to 4.
- [Release notes](https://github.com/actions/setup-node/releases)
- [Commits](https://github.com/actions/setup-node/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/setup-node
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump axios from 0.27.2 to 1.6.0 in /frontend (#7708)

* Bump axios from 0.27.2 to 1.6.0 in /frontend

Bumps [axios](https://github.com/axios/axios) from 0.27.2 to 1.6.0.
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/v1.x/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/v0.27.2...v1.6.0)

---
updated-dependencies:
- dependency-name: axios
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update NOTICES

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Vincent Donato <vincent@streamlit.io>

lukasmasuch committed:
Directly use package name in setup.py (#7717)

* Directly use package name in setup.py

* Fix update name script to replace correct name

snehankekre committed:
Remove unnecessary PIL dependency in st.image docstring (#7728)

LarsHill committed:
Allow to access external server IP via https. (#7712)

Fallback to HTTPS endpoint call to check external IP if HTTP call failed.

---------

Co-authored-by: Lars Hillebrand <lars.patrick.hillebrand@iais.fraunhofer.de>
Co-authored-by: Karen Javadyan <kajarenc@gmail.com>

vdonato committed:
Allow a service managing a Streamlit runtime to set its own session IDs (#7544)

eric-skydio committed:
Move `dg_stack` into a `ContextVar` to support `with` blocks in separate threads (#7715)

dependabot[bot] committed:
Bump actions/github-script from 6 to 7 (#7738)

Bumps [actions/github-script](https://github.com/actions/github-script) from 6 to 7.
- [Release notes](https://github.com/actions/github-script/releases)
- [Commits](https://github.com/actions/github-script/compare/v6...v7)

---
updated-dependencies:
- dependency-name: actions/github-script
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

kajarenc committed:
Improve st.connection caching behavior (#7730)

* hack to compute connection cache function keys based on ttl, max_entries and dsn

* Fix ttl and connection name behavior for all connections and factory

* Add explanatory comments

* More explanatory comment tweaks

* Avoid adding new `.` characters to function qualnames

---------

Co-authored-by: Vincent Donato <vincent@streamlit.io>

lukasmasuch committed:
Pin playwright version to < 1.40 (#7747)

* Pin playwright version to < 1.40

* Add comment

sfc-gh-tteixeira committed:
Use "loading skeletons" throughout Streamlit  (#7598)

* Implement both app-wide skeleton and suspense skeleton. Still need tests.

* Slow down the animation a little bit.

* Awww yisss. MFing tests.

* Address comments

* Make skeleton look nicer in sidebar

* Make DeckGlJsonChart withMapboxToken use Skeleton instead of "Loading..."

* Lint

* Fix assignment to const

* Add test for mapbox skeleton

* Make skeleton color overridable with postMessage

* Make a couple more e2e tests use AppSkeleton rather than "Please wait..."

* Make e2e tests not depend on skeletons. Instead, use new data-teststate attribute.

* Add query params for: no loading screen; old "Please wait" loading screen

* Fix AppNode test elements.alert check

* Lint

* Fix tests by creating a test-only "initial" state

* Fix e2e bug. "should not exist" -> "should exist"

* Pin playwright version to < 1.40 (#7747)

* Pin playwright version to < 1.40

* Add comment

* Add skeleton_background_color to NewSession.proto

* Fix arguments to cypress, whenever we call it manually.

* Add ability to pass function to matchThemedSnapshots()

* Ooops, remove Cypress upgrade code that was accidentally committed.

---------

Co-authored-by: Lukas Masuch <Lukas.Masuch@gmail.com>

lukasmasuch committed:
Migrate `st.button` e2e test to playwright (#7755)

* Add wait time

* Migrate button test to playwright

* Add snapshots

kajarenc committed:
Replace deprecated `find_executable` from `distutils.spawn` with recommended `which` from `shutil` (#7756)

Replace deprecated find_executable from distutils.spawn with recommended which from shutil
PEP: https://peps.python.org/pep-0632/

sfc-gh-pfinnigan committed:
add app_heartbeat BackMsg type and send it when receiving SEND_APP_HEARTBEAT window message (#7682)

lukasmasuch committed:
Remove outdated snapshots (#7758)

sfc-gh-dmatthews committed:
Update docstrings for release 1.29 (#7759)

* Style tweaks for connections docstrings

* Add border to container example

* Update data_editor docsring for timedelta addition

* Fix link format

* Clarify data_editor note re: timedelta

willhuang1997 committed:
Add ability to set query param key to tuple or set, but not dictionary

willhuang1997 committed:
Merge branch 'fix/bothQueryApisUsed' into query_params/acceptTuplesAndSets

samueldg committed:
Fix minor typo in server.scriptHealthCheckEnabled help (#7762)

vdonato commented on lib/streamlit/runtime/state/query_params.py:
```diff
@@ -49,8 +50,15 @@ def __getitem__(self, key: str) -> str:
+    def __setitem__(
+        self, key: str, value: Union[str, Set[str], Tuple[str], List[str]]
+    ) -> None:
+        if isinstance(value, dict):
+            raise StreamlitAPIException(
+                f"You cannot set a query params key `{key}` to a dictionary `{value}`."
+            )
```
I don't think we want to print the value of the dictionary here. The stacktrace + key name will already be more than enough for a developer to find the offending line of code, and the dict value is probably just noise

vdonato commented on lib/tests/streamlit/runtime/state/query_params_test.py:
```diff
@@ -67,12 +67,28 @@ def test__setitem__adds_list_value(self):
+        assert "test=1&test=2&test=3" in message.page_info_changed.query_string
+
+    def test__setitem__adds_set_value(self):
+        self.query_params["test"] = set((1, 2, 3))
```
can just write this as a set literal: `{1, 2, 3}`

ObservedObserver committed:
fix: component dev url typo (#7746)

BlackHC committed:
Don't use hash on floats in hashing.py (#7754)

* Don't use hash on floats

Python's hash is salted with a random number. This will not work across restarts. Instead, use struct.pack to get the bytes of a float.

* Run autoformatter

---------

Co-authored-by: Vincent Donato <vincent@streamlit.io>

vdonato committed:
Fix another url typo (#7764)

lukasmasuch committed:
Fix typing issues with altair (#7769)

* Fix typing issues with altair

* Fix mypy issues

* Fix mypy issues

* Remove altair from test dependencies

willhuang1997 committed:
Address comments

willhuang1997 committed:
Merge branch 'develop' into query_params/acceptTuplesAndSets
